### PR TITLE
Fix 'Argument list too long' error in get_remaining_jobs

### PR DESCRIPTION
**Pull request by tamohannes**
### The Issue

When running generation jobs with many random seeds and chunks (e.g., `--num_random_seeds 8 --num_chunks 120`), the pipeline would fail with:
```
bash: Argument list too long
```

This occurred in the `get_remaining_jobs` function in `nemo_skills/pipeline/utils/generation.py`, which checks which output files already exist to avoid reprocessing completed work.

#### Root Cause
The original implementation was concatenating all file existence checks into a single massive bash command:
```bash
bash -c 'if [ ! -f "file1" ]; then echo "MISSING:0:0"; fi; if [ ! -f "file2" ]; then echo "MISSING:0:1"; fi; ...'
```

With 960 files (8 seeds × 120 chunks) and long file paths containing lustre paths like `/lustre/fsw/llmservice_nemo_reasoning/...`, this created a command exceeding the system's maximum argument length limit (typically ~2MB).

### The Solution

The fix implements a simple, robust batching approach:

1. **Conservative Batching**: Process file checks in small batches of 30 files each, ensuring we stay well below system limits even with very long paths.

2. **Progress Logging**: For large file sets (>100 files), log progress as each batch is processed.

3. **Graceful Fallback**: If a batch fails, automatically fall back to checking files individually.

4. **No External Dependencies**: Avoids temporary files that caused issues with remote execution on Slurm clusters.

### Key Changes

```python
# Before: Flawed batching logic
request_size = len(check_commands[0]) // 10  # This made no sense
if len(random_seeds) > request_size:
    # Attempted to batch but logic was incorrect

# After: Simple, reliable batching
batch_size = 30  # Conservative size that works with long paths
for i in range(0, len(check_commands), batch_size):
    batch = check_commands[i : i + batch_size]
    # Process batch with proper error handling
```

### Benefits

- **Scalability**: Can handle any number of files without hitting system limits
- **Visibility**: Progress logging for large jobs (e.g., "Processing batch 1/32...")
- **Reliability**: Fallback mechanism ensures job completion even if some batches fail
- **Simplicity**: Straightforward implementation that's easy to understand and maintain

### Testing

The fix was tested with the failing configuration (8 random seeds × 120 chunks = 960 files) and successfully processed all file checks without errors.

### Code Diff

The main changes are in `nemo_skills/pipeline/utils/generation.py`:
- Removed the flawed batching logic based on command length
- Removed temporary file approach that failed with remote execution
- Implemented simple batch processing with size 30
- Added progress logging for better user experience
- Enhanced error handling with per-file fallback

This fix ensures that NeMo-Skills can reliably handle large-scale generation jobs regardless of the number of random seeds or chunks specified.

### CODE REVIEW

File: nemo_skills/pipeline/utils/generation.py [hunk 1]
```diff
@@ -188,7 +211,7 @@ def get_generation_cmd(
         cmd += (
             f"    ++inference.random_seed={random_seed} "
             f"    ++inference.temperature=0.7 "
-            f"    ++inference.top_k=-1 "
+            f"    ++inference.top_k=0 "
```
Comment (Kipok) [inline nemo_skills/pipeline/utils/generation.py]: please roll-back unrelated change

File: nemo_skills/pipeline/utils/generation.py [hunk 2]
```diff
@@ -318,7 +341,7 @@ def configure_client(
         }
         extra_arguments = (
             f"{extra_arguments} ++server.server_type={server_type} "
-            f"++server.host=localhost ++server.port={server_port} ++server.model={model} "
+            f"++server.host=localhost ++server.port={server_port} "
```
Comment (Kipok) [inline nemo_skills/pipeline/utils/generation.py]: please roll-back unrelated change

File: nemo_skills/pipeline/utils/generation.py [hunk 3]
```diff
@@ -74,25 +74,48 @@ def get_remaining_jobs(cluster_config, output_dir, random_seeds, chunk_ids, reru
...
+                single_command = f"bash -c '{cmd}'"
+                try:
+                    if cluster_config['executor'] == 'slurm':
+                        out = get_tunnel(cluster_config).run(single_command).stdout.strip()
+                    else:
+                        out = subprocess.run(single_command, shell=True, check=True, stdout=subprocess.PIPE).stdout.decode("utf-8")
+                    if out:
+                        outputs.append(out)
+                except Exception as inner_e:
+                    LOG.error(f"Failed to check file {i+j+1}/{total_files}: {inner_e}")
```
Comment (Kipok) [inline nemo_skills/pipeline/utils/generation.py]: we should fail with error in this case, right?

Comment (tamohannes) [reply nemo_skills/pipeline/utils/generation.py]: Yes, exactly, I've just made the changes 🙌🏻

**Review by Kipok — APPROVED** [review]
looks good to me as long as this is tested well. Please make sure to try it both locally and on slurm, with / without chunks and in a few different configurations. Ideally we should add ci tests for this, but can at least test manually for now

**Comment by tamohannes** [comment]
Hi @Kipok,  

Thanks for the review! I've addressed all your comments 🚀:

1. **Rolled back unrelated changes**:  
   - Reverted `top_k` from `0` back to `-1`  
   - Restored the `++server.model={model}` parameter  

2. **Added proper error handling**:  
   - The code now raises `RuntimeError` when individual file checks fail, as requested  

3. **Added comprehensive CI tests** (`tests/test_pipeline_utils.py`):  
   - Tests for small batches (no batching needed)  
   - Tests for large batches (960 files — the original failing case)  
   - Tests for very large batches (2000+ files)  
   - Tests for batch failure fallback behavior  
   - Tests for both local and Slurm execution  
   - Tests for various edge cases (no chunks, `rerun_done` flag, etc.)  

4. **Tested various configurations**:  
   - The new tests cover different combinations of random seeds and chunks  
   - Includes tests for both local and Slurm cluster configurations  
   - All 8 tests are passing  

The CI tests follow the project's conventions and style, ensuring consistency and maintainability.
Let me know if you spot anything else that needs tweaking. 🙏🏻
